common/dal2/src/ops/io.rs

```diff
@@ -12,10 +12,72 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::io;
 use std::pin::Pin;
+use std::task::Context;
 use std::task::Poll;
 
-pub type Reader = Box<dyn futures::io::AsyncRead + Unpin + Send>;
+use bytes;
+use futures;
+use futures::ready;
+use futures::AsyncRead;
+use pin_project::pin_project;
+
+pub type Reader = Box<dyn AsyncRead + Unpin + Send>;
+
+const CAPACITY: usize = 4096;
+
+/// ReaderStream is used to convert a `futures::io::AsyncRead` into a `futures::Stream`.
+///
+/// Most code inspired by `tokio_util::io::ReaderStream`.
+#[pin_project]
+pub struct ReaderStream {
+    #[pin]
+    reader: Option<Reader>,
+    buf: bytes::BytesMut,
+}
+
+impl ReaderStream {
+    pub fn new(r: Reader) -> Self {
+        ReaderStream {
+            reader: Some(r),
+            buf: bytes::BytesMut::new(),
+        }
+    }
+}
+
+impl futures::Stream for ReaderStream {
+    type Item = Result<bytes::Bytes, io::Error>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.as_mut().project();
+
+        let mut reader = match this.reader.as_pin_mut() {
+            Some(r) => r,
+            None => return Poll::Ready(None),
+        };
+
+        // We will always use the same underlying buffer, the allocation happens only once.
+        if this.buf.is_empty() {
+            this.buf.resize(CAPACITY, 0);
+        }
+
+        match ready!(reader.as_mut().poll_read(cx, this.buf)) {
```

Comment:
do we need the `as_mut()` here?  would `ready!(reader.poll_read...` be  ok as well?

Reply:
Yep, addressed.